use a try block in _expand_user instead of a typing condition

neutrinoceros opened the pull request:
Noticed that `_expand_user`'s docstrings stated that expansion was performed "if possible", while the actual implementation was only applying expansion to str objects.
With a try block it should work with str, os.Pathlike and bytes objects (and everything else supported by `os.path.expanduser` in the future).
I except this change may be undesirable since it will force the output type to `str`. In this case I can replace this with a docstring update.
I'll complete the checklist with tests and whatsnew entry if the change is deemed desirable.

- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

neutrinoceros committed:
refactor: use a try block in _expand_user instead of a typing condition. Net effect: '~' token expansion affects paths objects too.

arw2019 reviewed (changes requested):
others need to weigh in on whether we want to make this change

but if we do go ahead we'll need:
- [ ] tests showing the new behaviour /modify existing tests. Look in `pandas/tests/io/test_common.py`
- [ ] a docstring update, as you said
- [ ] a release note (this probably goes to 1.2)

neutrinoceros commented:
@arw2019 thanks, will do!
I gave up on this when I couldn't figure out how to make mypy happy with the change, any piece of advice here ?

arw2019 commented:
> @arw2019 thanks, will do!
> I gave up on this when I couldn't figure out how to make mypy happy with the change, any piece of advice here ?

Hmmm it seems like there are a ton of CI failures rn. Can you rebase against master and see which problems persist

neutrinoceros commented:
> Can you rebase against master and see which problems persist

Done !
I note the logic block I'm touching has been updated on master since I opened this, but I still think a try-except pattern is probably more robust than what's already there. I'll be back to add tests to this :)

neutrinoceros committed:
fix a monkeypatch for _expand_user in test

neutrinoceros committed:
raise from

neutrinoceros committed:
ignore typing checks in tryblock

neutrinoceros commented:
Think I fixed the branch now !

arw2019 commented:
> Think I fixed the branch now !

Ok, looping in some of the core team to see if there's appetite for this change

@simonjayhawkins @jreback @mroeschke

neutrinoceros committed:
add tests

pep8speaks commented:
Hello @neutrinoceros! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/io/test_common.py`](https://github.com/pandas-dev/pandas/blob/463aaa185639d71d065d92b63fe3623b1141d29f/pandas/tests/io/test_common.py):

> [Line 86:1](https://github.com/pandas-dev/pandas/blob/463aaa185639d71d065d92b63fe3623b1141d29f/pandas/tests/io/test_common.py#L86): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

neutrinoceros commented:
I've now added tests to check for the new behaviour. The new tests fail on master and pass on this branch.

jbrockmendel commented on pandas/io/common.py:
```diff
@@ -243,12 +228,19 @@ def get_filepath_or_buffer(
+    except TypeError as e:
         msg = f"Invalid file path or buffer object type: {type(filepath_or_buffer)}"
-        raise ValueError(msg)
+        raise ValueError(msg) from e
```
can you use "err" instead of "e"?  1-letter variable names are hard to grep for

neutrinoceros replied:
done !

neutrinoceros committed:
avoid single-char variable name

dsaxton commented:
@neutrinoceros Is this still active? Can you merge master to fix conflicts if so?

neutrinoceros commented:
Hi ! Yes I am still willing to go through with this. I just rebased in a pretty agressive way, maybe I'll need to do some back pedalling. I see the parts I changed in this PR have seen considerable churn since I originally read the code so maybe the added value is gone now.

jreback commented on pandas/io/common.py:
```diff
@@ -322,18 +309,19 @@ def get_filepath_or_buffer(
-            should_close=False,
-            mode=mode,
-        )
+    if is_file_like(filepath_or_buffer) or isinstance(filepath_or_buffer, mmap.mmap):
```
can we change the definition of is_file_like to *include* mmamp.mmap (e.g. does this break anything generally)?

neutrinoceros replied:
since mmap.mmap objects have read/write methods and are iterables, it seems very much in line with the docstring in `is_file_like`, so I think this is an excellent idea.

jreback commented on pandas/io/common.py:
```diff
@@ -322,18 +309,19 @@ def get_filepath_or_buffer(
+            compression,
+            False,
+        )
+    except TypeError as err:
```
instead of this try/except (and the fallback arg), can we just raise this inside _expand_user (and have this buble up)?

neutrinoceros commented:
Ok I must admit I'm having a hard time trying to put the pieces back together and I am not sure this is worth it.
I want thanks @jreback for taking the time to review this, and everyone involved in the discussion, but I now think the easiest way to fix this would be to just start fresh. Maybe I'll reissue this later, but I don't wish to clutter the issue tracker.
